network/stream/stream.go

```diff
@@ -1131,21 +1098,19 @@ func (r *Registry) Stop() error {
 	log.Debug("stream registry stopping")
 	r.mtx.Lock()
 	defer r.mtx.Unlock()
-
 	close(r.quit)
-	// wait for all handlers to finish
-	done := make(chan struct{})
-	go func() {
-		r.handlersWg.Wait()
-		close(done)
-	}()
-	select {
-	case <-done:
-	case <-time.After(5 * time.Second):
+
```

Comment:
This shutdown was/is a bit weird :) If I understood correctly, the old code closed the `quit` channel in registry and was basically looping the messages without executing the  async task. It did not stop the loop at this point. For every execution of the async part, it was increasing the wg handle. Then, on shutdown it was waiting for the wg handle for 5 seconds, to gracefully shutdown, while still relying on some peer.Drop to happen after to cancel to run loop.
In this PR I  have removed the wg handle from the registry, and I moved the shutdown and wait logic into the peer. So, now registry needs to range over all peers in initiate the shutdown and wait for them to wait or timeout, which results in the similar behavior, but a more simple handler, imo. This way, the `quit` channel is not used to actually stop the async part, but to make the handler logic quit appropriately, and the actual Peer.Shutdown is stopping the execution of the async jobs, and also returning `nil` from the run loop which should result in disconnecting the peer, so it shift a bit more control into the protocol.Run instead of relying on side effect. At least this was my rationale. Does it makes sense to you as well? :)

Reply:
makes sense